### PR TITLE
[MachineCopyPropagation] When the source of PreviousCopy is undef, we cannot replace sub register

### LINKED ISSUE #74680: Wrong register used on apple-m1
I tried this IR/MIR:

```llvm
; ModuleID = 'repro_117902.54923e0577e9cdaf-cgu.0'
source_filename = "repro_117902.54923e0577e9cdaf-cgu.0"
target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
target triple = "arm64-apple-macosx11.0.0"

define i1 @foo(i32 %self.0.val, i8 %self.4.val) #1 {
start:
  switch i32 %self.0.val, label %bb2 [
    i32 0, label %bb9
    i32 1, label %bb4
    i32 2, label %bb5
    i32 3, label %bb1
  ]

bb2:                                              ; preds = %start
  unreachable

bb4:                                              ; preds = %start
  br label %bb9

bb5:                                              ; preds = %start
  br label %bb9

bb1:                                              ; preds = %start
  %trunc.not = icmp eq i8 %self.4.val, 0
  br label %bb9

bb9:                                              ; preds = %bb1, %start, %bb5, %bb4
  %r = phi i1 [ true, %bb5 ], [ true, %bb4 ], [ false, %start ], [ %trunc.not, %bb1 ]
  ret i1 %r
}

attributes #1 = { "frame-pointer"="non-leaf" "target-cpu"="apple-m1" }
```

I got:

```
_foo:                                   ; @foo
        mov     x8, x0
        mov     w0, #0                          ; =0x0
        adrp    x9, LJTI0_0@PAGE
        add     x9, x9, LJTI0_0@PAGEOFF
        adr     x10, LBB0_1
        ldrb    w11, [x9, x8]
        add     x10, x10, x11, lsl #2
        br      x10
...
```

If I run with `-O1`:

```
_foo:                                   ; @foo
        mov     x8, x0
        mov     w8, w0
        mov     w0, #0                          ; =0x0
        adrp    x9, LJTI0_0@PAGE
        add     x9, x9, LJTI0_0@PAGEOFF
        adr     x10, LBB0_1
        ldrb    w11, [x9, x8]
        add     x10, x10, x11, lsl #2
        br      x10
```

I think the first instruction should be `mov w8, w0`.

I'm not familiar with the MIR. But it looks like postrapseudos created `mov x8, x0` in lowercopy.
Then machine-cp removes late `mov w8, w0`.

godbolt: https://llvm.godbolt.org/z/1hfYvrGPG
from: https://github.com/rust-lang/rust/issues/117902

### PR DESCRIPTION
**Pull request by dianqk**
Fixes #74680.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/machine-cp-undef.mir [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# RUN: llc -o - %s -O3 -mcpu=apple-m1 -mtriple=arm64-apple-macos | FileCheck %s
```
Comment (nikic) [inline llvm/test/CodeGen/AArch64/machine-cp-undef.mir:1]: Can this be reduced to just a `-run-pass=machine-cp` test? Please also update update_mir_checks.py.

Comment (DianQK) [reply llvm/test/CodeGen/AArch64/machine-cp-undef.mir:1]: > Can this be reduced to just a `-run-pass=machine-cp` test? Please also update update_mir_checks.py.

I think it's possible, but I'm still trying. After the reduction is complete, I'll switch to update_mir_checks.py.

Comment (DianQK) [reply llvm/test/CodeGen/AArch64/machine-cp-undef.mir:1]: Interestingly, I can use it in with `--start-before`.
First run with `--start-before=machine-cp,0 --stop-before=machine-cp,2`, then run with `--start-before=machine-cp,2 --stop-after=machine-cp,2`. The output is:

```
MCP: Copy is a deletion candidate:   $x8 = ORRXrs $xzr, undef $x0, 0, implicit $w0
MCP: copy is a NOP, removing:   dead renamable $w8 = ORRWrs $wzr, $w0, 0, implicit-def dead $x8
```

Maybe an analysis pass is missing?

Comment (DianQK) [reply llvm/test/CodeGen/AArch64/machine-cp-undef.mir:1]: Resolved. I need add `-mcp-use-is-copy-instr`.

**Comment by DianQK** [comment]
> Hello. I think that if you removed undef from the first instruction the result would still be incorrect. With:
> 
> ```
> $x8 = ORRXrs $xzr, $x0, 0, implicit $w0
> $w8 = ORRWrs $wzr, $w0, 0, implicit-def $x8
> ```

I'm also curious about it, but this transformation has been around for a long time, so I'm assuming it's correct. I think there exists some other pass solution to this problem here.

> The top bits are not zero into the function?

Yes. Please see the link https://llvm.godbolt.org/z/zv3vz8zcs.

```asm
_call_from_main:
        sub     sp, sp, #32
        stp     x29, x30, [sp, #16]
        add     x29, sp, #16
        mov     w8, #1
        str     w8, [sp, #4] ;;; We save `0x00000001` on stack
        add     x0, sp, #4 ;;; Save the address
        bl      _device_create_texture
        ldp     x29, x30, [sp, #16]
        add     sp, sp, #32
        ret
_device_create_texture:
        sub     sp, sp, #32
        stp     x29, x30, [sp, #16]
        add     x29, sp, #16
        ldr     x8, [x0] ;;; Load 8-byte, get `0x6f60400000000001`
        str     x8, [sp]
        ldr     w9, [x0, #8]
        str     w9, [sp, #8]
        ubfx    x1, x8, #32, #8
        mov     x0, x8
        bl      __ZN12repro_11790213TextureFormat17required_features17h5e4d5de64322154bE
        ..
        ret
__ZN12repro_11790213TextureFormat17required_features17h5e4d5de64322154bE:
        mov     x8, x0 ;;;  x8/x0 is `0x6f60400000000001`
        mov     w0, #0
        adrp    x9, LJTI10_0@PAGE
        add     x9, x9, LJTI10_0@PAGEOFF
        adr     x10, LBB10_1
        ldrb    w11, [x9, x8] ;;; We expect x8 is `0x0000000000000001`
```

In this function, we can't guarantee that `x8/x0` is 0x1, we have to do a transformation `mov w8, w8`.
So I think this delete subregister copy instruction is based on a convention similar to this one. We can delete `mov w8, w0` on this:

```asm
mov x8, x0
mov w8, w0
mov w8, w8
```

If I use `-O0`, I can get the result:

```asm
__ZN12repro_11790213TextureFormat17required_features17h5e4d5de64322154bE:
        sub     sp, sp, #16
        str     w1, [sp, #4]
        subs    w8, w0, #0
        mov     w8, w8 ; This looks like a transformation that guarantees that x8 is `0x0000000000000001`.
```

After that I found the following code in `lowerCopy`.

https://github.com/llvm/llvm-project/blob/0808be47b8fbf0307d0b6f2eb45ba9bfe1b3ae65/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L4422-L4429

If we lower the code to `mov x8, x0` instead of `mov w8, w0`, this may be somewhat unsafe. So this code leaves an undef.
My PR was answering this comment and undef.

**Comment by davemgreen** [comment]
I don't believe the undef is the issue - I think the issue is that AArch64InstrInfo::isCopyInstrImpl is saying that a W-reg orr is a copy, even if it is really a zextend because the entire X output register is depended upon.

Can you try and add something to isCopyInstImpl instead, that says: If the register is virtual then it should not be a subreg, and if the register is physical then there should not be an implicit def of the X reg. Would that solve your issue, and would it cause other problems in AArch64 codegen? Thanks

**Comment by DianQK** [comment]
> I don't believe the undef is the issue - I think the issue is that AArch64InstrInfo::isCopyInstrImpl is saying that a W-reg orr is a copy, even if it is really a zextend because the entire X output register is depended upon.

Thanks for the explanation.

> Can you try and add something to isCopyInstrImpl instead, that says: If the register is virtual then it should not be a subreg, and if the register is physical then there should not be an implicit def of the X reg. Would that solve your issue, and would it cause other problems in AArch64 codegen? Thanks

I'm not sure if I'm misunderstanding anything. But I tried to file the #75184 PR.

It breaks the original commit https://github.com/llvm/llvm-project/commit/53726588f672a915c6f907569356091552500f23. It looks like it's because we're generating the mov instruction for the ORRWsr. (At this point this is a copy instruction?) I'm not sure exactly what is going wrong. Maybe we need to add a new method to TargetRegisterInfo?

Or are you going to fix this issue? I'm not familiar with MIR.
Alternatively, we can add implicit-def to MachineCopyPropagation? But that doesn't seem right either.

**Comment by davemgreen** [comment]
Thanks. It sounds like there are not a lot of code changes, which is a good sign. I didn't expect the debug problems though.

I'll try and take a look at the patch. Perhaps you are right that we need a new method for the debug info to use.

**Comment by DianQK** [comment]
> I'll try and take a look at the patch. Perhaps you are right that we need a new method for the debug info to use.

Based on https://github.com/llvm/llvm-project/blob/fd8fa31c55a3413f643443ecf3301441428ce513/llvm/docs/InstrRefDebugInfo.md?plain=1#L125-L127, perhaps we could add an `isCopyLikeInstr` method? This should be a good name while changing very little.I 'll give it a try later.

**Comment by DianQK** [comment]
I tried adding `isCopyLikeInstr` to #75184. All known test cases have passed.

**Pull request closed without merging**